### DISCUSSION TITLE
feature request: ask for permission on epoptes client

### DISCUSSION THREAD

**Discussion by jankowa** [epoptes/epoptes#67] [Q&A]
Hi,

we use ltsp a lot in office environments. Epoptes would also be a very nice tool for helpdesk. But since the default is that an administrator could look at any desktop without even notify the user, this would be an privacy issue.

Even in educational context I would prefer that everybody know's if he gets controlled or not.

I didn't found any possibility in configuration to force epoptes / x11vnc asking for permissions in a user session.

The easiest way would be some zenity query or something like below. Maybe configurable in /etc/default/epoptes-client:

So if a MONITOR_MSG /ASSIST_MSG exists in /etc/default/epoptes-client it will ask for confirmation.

Of course it's easy to bypass this. But this could be achieved in any case by the admin. But it feels much better, if user and administrators know that there is any kind of confirmation before someone looks at a users desktop.

A quick hack (with zenity as dependency):

```
--- client-functions.orig	2018-09-12 11:50:21.000000000 +0200
+++ client-functions	2018-09-12 15:04:25.597658561 +0200
@@ -307,11 +307,23 @@
     background ./message "$1" "$2" "$3"
 }
 
+if [ -f /etc/default/epoptes-client ]; then
+    . /etc/default/epoptes-client
+fi
+
+if [ "$MONITOR_MSG" ]; then
+    ACCEPT_STR_MONITOR="zenity --question --text '${MONITOR_MSG}'"
+fi
+
+if [ "$ASSIST_MSG" ]; then
+    ACCEPT_STR_ASSIST="zenity --question --text '${ASSIST_MSG}'"
+fi
+
 # Connect to the server to be monitored.
 # Parameters:
 # $1 = port.
 get_monitored() {
-    background x11vnc -noshm -24to32 -viewonly -connect_or_exit "$SERVER:$1"
+    background x11vnc -noshm -24to32 -accept "${ACCEPT_STR_MONITOR}" -viewonly -connect_or_exit "$SERVER:$1"
 }
 
 # Connect to the server to get assistance.
@@ -319,7 +331,7 @@
 # $1 = port.
 # $2 = grab keyboard and mouse.
 get_assisted() {
-    background x11vnc -noshm -24to32 ${2:+-grabptr -grabkbd} -connect_or_exit "$SERVER:$1"
+    background x11vnc -noshm -24to32 ${2:+-grabptr -grabkbd} -accept "${ACCEPT_STR_ASSIST}" -connect_or_exit "$SERVER:$1"
 }
 
 # Deactivate the screensaver, in order for the users to watch a broadcast.


```

In /etc/default/epoptes-client something like
```

MONITOR_MSG="Confirm for remote session..."
ASSIST_MSG="Confirm for remote session with remote control ...."
```

**Comment by alkisg** [comment]
If confirmation is required, why not have the users use the remote assistance feature instead?
This way the connection is user-initiated, they don't need to confirm anything.

**Comment by jankowa** [comment]
Well, epoptes is fine for e.g. starting root console, seeing what machines are up and so on. But while installed epoptes it would be also possible to watch screens without user notice about. No matter if admins normally use x11vnc.

Of course I can use plain x11vnc user initiated helpdesk (and I use).  But without the convenience of epoptes. And to be honest: for some users it's even hard to understand the x11vnc interface. A simple "ok" is much easier.

**Comment by alkisg** [comment]
I'll close this issue as it can be easily implemented by a wrapper script in /usr/local/bin/x11vnc on the clients, which invokes x11vnc with modified parameters, and not the ones passed by Epoptes.

* I.e. Epoptes would run: `background x11vnc -noshm -24to32 ${2:+-grabptr -grabkbd} -connect_or_exit "$SERVER:$1"`
* That would launch your script at /usr/local/bin/x11vnc
* Your script would run: `exec /usr/bin/x11vnc -accept "${ACCEPT_STR_ASSIST}" "$@"`

**Comment by jankowa** [comment]
Great, thanks for the idea. This is indeed much simpler than hacking epoptes itself.
